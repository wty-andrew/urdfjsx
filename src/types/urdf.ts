import { JOINT_TYPES } from '../constants/index.js'
import type { RequiredPick } from './common.js'

export type Vector3 = [number, number, number]

export type Vector4 = [number, number, number, number]

export type Color = Vector4

export type Origin = {
  xyz?: Vector3
  rpy?: Vector3
}

export type Limit = {
  upper: number
  lower: number
}

export type JointType = (typeof JOINT_TYPES)[number]

export type Joint = {
  name: string
  type: JointType
  origin?: Origin
  parent: string
  child: string
  axis?: Vector3
  limit?: Limit
}

export type Material = {
  name?: string
  color?: Color
  texture?: string
}

export type KnownMaterial =
  | RequiredPick<Material, 'name'>
  | RequiredPick<Material, 'color'>

export type Box = {
  type: 'box'
  size: Vector3
}

export type Cylinder = {
  type: 'cylinder'
  radius: number
  length: number
}

export type Sphere = {
  type: 'sphere'
  radius: number
}

export type Mesh = {
  type: 'mesh'
  filename: string
  scale?: Vector3
}

export type Geometry = Box | Cylinder | Sphere | Mesh

export type Visual = {
  origin?: Origin
  geometry: Geometry
  material?: Material
}

export type Collision = {
  origin?: Origin
  geometry: Geometry
}

export type Link = {
  name: string
  type?: string
  visual: Visual[]
  collision: Collision[]
}

export type Robot = {
  name: string
  link: Link[]
  joint: Joint[]
  material: Material[]
}
