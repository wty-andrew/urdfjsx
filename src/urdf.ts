import * as R from 'ramda'

import type { Dict, NestedArray, RequiredPick } from './types.js'
import { isString, parseNumbers } from './utils.js'
import { Element, parseXML } from './xml.js'

// only process a subset of urdf elements, see http://wiki.ros.org/urdf/XML for full spec
const CHILD_ELEMENTS: Dict<string[]> = {
  robot: ['link', 'joint', 'material'],
  link: ['visual', 'collision'],
  joint: ['origin', 'parent', 'child', 'axis', 'limit'],
  visual: ['origin', 'geometry', 'material'],
  collision: ['origin', 'geometry'],
  geometry: ['box', 'cylinder', 'sphere', 'mesh'],
  material: ['color', 'texture'],
}

const JOINT_TYPES = [
  'revolute',
  'continuous',
  'prismatic',
  'fixed',
  'floating',
  'planar',
] as const

export type JointType = (typeof JOINT_TYPES)[number]

export type Vector3 = [number, number, number]

export type Vector4 = [number, number, number, number]

export type Color = Vector4

export type Origin = {
  xyz?: Vector3
  rpy?: Vector3
}

export type Limit = {
  upper: number
  lower: number
}

export type Joint = {
  name: string
  type: JointType
  origin?: Origin
  parent: string
  child: string
  axis?: Vector3
  limit?: Limit
}

export type Material = {
  name?: string
  color?: Color
  texture?: string
}

export type KnownMaterial =
  | RequiredPick<Material, 'name'>
  | RequiredPick<Material, 'color'>

export type Box = {
  type: 'box'
  size: Vector3
}

export type Cylinder = {
  type: 'cylinder'
  radius: number
  length: number
}

export type Sphere = {
  type: 'sphere'
  radius: number
}

export type Mesh = {
  type: 'mesh'
  filename: string
  scale?: Vector3
}

export type Geometry = Box | Cylinder | Sphere | Mesh

export type Visual = {
  origin?: Origin
  geometry: Geometry
  material?: Material
}

export type Collision = {
  origin?: Origin
  geometry: Geometry
}

export type Link = {
  name: string
  type?: string
  visual: Visual[]
  collision: Collision[]
}

export type Robot = {
  name: string
  link: Link[]
  joint: Joint[]
  material: Material[]
}

//

type Result<T = any> =
  | { success: true; value: T }
  | { success: false; errors: NestedArray<string> }

type Extractor<T = any> = (el: Element) => Result<T>

const failure = (errors: NestedArray<string>): Result => ({
  success: false,
  errors,
})

const success = <T>(value: T): Result<T> => ({ success: true, value })

const string = (text?: string): Result<string> =>
  text === undefined ? failure(['Missing']) : success(text)

const float = (text?: string): Result<number> => {
  if (text === undefined) return failure(['Missing'])

  try {
    return success(parseFloat(text))
  } catch (error) {
    return failure([`Invalid number: "${text}"`])
  }
}

const vector3 = (text?: string): Result<Vector3> => {
  if (text === undefined) return failure(['Missing'])

  try {
    const numbers = parseNumbers(text)
    if (numbers.length !== 3) throw new Error()
    return success(numbers as Vector3)
  } catch (error) {
    return failure([`Invalid vector3: "${text}"`])
  }
}

const vector4 = (text?: string): Result<Vector4> => {
  if (text === undefined) return failure(['Missing'])

  try {
    const numbers = parseNumbers(text)
    if (numbers.length !== 4) throw new Error()
    return success(numbers as Vector4)
  } catch (error) {
    return failure([`Invalid vector4: "${text}"`])
  }
}

const jointType = (text?: string): Result<JointType> => {
  if (text === undefined) return failure(['Missing'])

  return JOINT_TYPES.includes(text as any)
    ? success(text)
    : failure([`Invalid joint type: "${text}"`])
}

const tag =
  <T extends string, U extends string>(
    obj: Record<T, U>
  ): Extractor<Record<T, U>> =>
  (el: Element) => {
    const entries = Object.entries(obj)
    if (entries.length !== 1) {
      throw new Error(`Invalid argument: "${JSON.stringify(obj)}"`)
    }

    return el.tag === entries[0][1]
      ? success(obj)
      : failure([`Expected tag: ${entries[0][1]}, received: ${el.tag}`])
  }

const attribute =
  <T extends Dict<(text: string) => Result>>(
    obj: T
  ): Extractor<{
    [K in keyof T]: T[K] extends (attr: string) => Result<infer R> ? R : string
  }> =>
  (element: Element) =>
    Object.entries(obj).reduce((accResult, [name, extractor]) => {
      if (!accResult.success) return accResult

      const result = extractor(element.attrib[name])
      return result.success
        ? success({ ...accResult.value, [name]: result.value })
        : failure([`Invalid attribute: "${name}"`, result.errors])
    }, success({})) as Result

const only =
  <T extends object | any[]>(
    extractor: Extractor<T>
  ): Extractor<T extends Array<infer R> ? R : T[keyof T]> =>
  (el: Element) => {
    const result = extractor(el)
    if (!result.success) return result
    const values = Array.isArray(result.value)
      ? result.value
      : Object.values(result.value)
    return values.length === 1
      ? success(values[0])
      : failure([`Expect single item, received: ${values}`])
  }

const children =
  <T extends Dict<(elements: Element[]) => Result>>(
    obj: T
  ): Extractor<{
    [K in keyof T]: T[K] extends (elements: Element[]) => Result<infer R>
      ? R
      : never
  }> =>
  (element: Element) => {
    const groups = R.groupBy(R.prop('tag'), element.children)
    return Object.entries(obj).reduce((accResult, [name, extractor]) => {
      if (!accResult.success) return accResult

      const result = extractor(groups[name] || [])
      return result.success
        ? success({ ...accResult.value, [name]: result.value })
        : failure([`Invalid ${element.tag} children: ${name}`, result.errors])
    }, success({})) as Result
  }

const child =
  <T>(extractor: Extractor<T>): Extractor<T> =>
  ({ tag, children }: Element) =>
    children.length === 1
      ? extractor(children[0])
      : failure([
          `Expect ${tag} element to have single child, received: ${children.length}`,
        ])

function combine<T>(e: Extractor<T>): Extractor<T>

function combine<T1, T2>(
  e1: Extractor<T1>,
  e2: Extractor<T2>
): Extractor<T1 & T2>

function combine<T1, T2, T3>(
  e1: Extractor<T1>,
  e2: Extractor<T2>,
  e3: Extractor<T3>
): Extractor<T1 & T2 & T3>

function combine<T extends Extractor<Dict<any>>[]>(...extractors: T) {
  return (el: Element) =>
    extractors.reduce((accResult, extractor) => {
      if (!accResult.success) return accResult
      const result = extractor(el)
      return result.success
        ? success({ ...accResult.value, ...result.value })
        : result
    }, success({}))
}

function optional<T>(
  extractor: (text: string) => Result<T>
): (text?: string) => Result<T | undefined>

function optional<T>(
  extractor: Extractor<T>
): (elements?: Element[]) => Result<T | undefined>

function optional<T>(extractor: any) {
  return (arg?: Element[] | string): Result<T | undefined> => {
    if (arg === undefined) return success(undefined)
    if (isString(arg)) return extractor(arg)
    if (arg.length === 0) return success(undefined)
    return arg.length > 1
      ? failure([
          `Expect elements to have length 0 or 1, received: ${arg.length}`,
        ])
      : extractor(arg[0])
  }
}

const one =
  <T>(extractor: Extractor<T>) =>
  (elements: Element[]): Result<T> =>
    elements.length === 1
      ? extractor(elements[0])
      : failure([
          `Expect elements to have length 1, received: ${elements.length}`,
        ])

const many =
  <T>(extractor: Extractor<T>) =>
  (elements?: Element[]): Result<T[]> =>
    elements === undefined
      ? success([])
      : elements.reduce(
          (accResult, element) => {
            if (!accResult.success) return accResult
            const result = extractor(element)
            return result.success
              ? success([...accResult.value, result.value])
              : result
          },
          success([] as T[])
        )

const oneOf =
  <T extends Extractor[]>(
    ...extractors: T
  ): Extractor<T[number] extends Extractor<infer P> ? P : never> =>
  (el: Element) => {
    const finalResult = extractors.reduce((accResult, extractor) => {
      if (accResult.success) return accResult
      const result = extractor(el)
      return result.success
        ? result
        : failure([...accResult.errors, result.errors])
    }, failure([]))

    return finalResult.success
      ? finalResult
      : failure(['No match', ...finalResult.errors])
  }

// mostly follow https://github.com/ros/urdfdom/blob/master/xsd/urdf.xsd

const origin: Extractor<Origin> = attribute({
  xyz: optional(vector3),
  rpy: optional(vector3),
})

const axis = only(attribute({ xyz: vector3 }))

const limit: Extractor<Limit> = attribute({ lower: float, upper: float })

const joint: Extractor<Joint> = combine(
  attribute({ name: string, type: jointType }),
  children({
    origin: optional(origin),
    parent: one(only(attribute({ link: string }))),
    child: one(only(attribute({ link: string }))),
    axis: optional(axis),
    limit: optional(limit),
  })
)

const color = only(attribute({ rgba: vector4 }))

const texture = only(attribute({ filename: string }))

const material: Extractor<Material> = combine(
  attribute({ name: optional(string) }),
  children({
    color: optional(color),
    texture: optional(texture),
  })
)

const box: Extractor<Box> = combine(
  tag({ type: 'box' }),
  attribute({ size: vector3 })
)

const cylinder: Extractor<Cylinder> = combine(
  tag({ type: 'cylinder' }),
  attribute({
    radius: float,
    length: float,
  })
)

const sphere: Extractor<Sphere> = combine(
  tag({ type: 'sphere' }),
  attribute({ radius: float })
)

const mesh: Extractor<Mesh> = combine(
  tag({ type: 'mesh' }),
  attribute({
    filename: string,
    scale: optional(vector3),
  })
)

const geometry = child(oneOf(box, cylinder, sphere, mesh))

const visual: Extractor<Visual> = children({
  origin: optional(origin),
  geometry: one(geometry),
  material: optional(material),
})

const collision: Extractor<Collision> = children({
  origin: optional(origin),
  geometry: one(geometry),
})

const link: Extractor<Link> = combine(
  attribute({ name: string, type: optional(string) }),
  children({
    visual: many(visual),
    collision: many(collision),
  })
)

const robot: Extractor<Robot> = combine(
  attribute({ name: string }),
  children({
    link: many(link),
    joint: many(joint),
    material: many(material),
  })
)

//

const prune = (element: Element): Element => {
  const { tag, attrib, children } = element
  return {
    tag,
    attrib,
    children: children
      .filter((child) => CHILD_ELEMENTS[tag].includes(child.tag))
      .map(prune),
  }
}

const formatError = (errors: NestedArray<string>, indent = 0): string => {
  if (errors.length === 0) return ''
  const [first, ...rest] = errors
  return (
    (isString(first)
      ? `\n${'  '.repeat(indent)}${first}`
      : formatError(first, indent + 1)) + formatError(rest, indent)
  )
}

export const parseURDF = (text: string) => {
  const result = robot(prune(parseXML(text)[0]))
  if (!result.success) {
    throw new Error(formatError(result.errors))
  }
  return result.value
}

export const getRootLink = (robot: Robot) => {
  const isChildLink = R.includes(R.__, R.pluck('child', robot.joint))
  const rootLinks = R.reject(isChildLink, R.pluck('name', robot.link))
  if (rootLinks.length !== 1) {
    throw new Error('URDF must have only one root')
  }
  return rootLinks[0]
}
