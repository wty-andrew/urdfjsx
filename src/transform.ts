import * as t from '@babel/types'
import * as R from 'ramda'

import type { Dict } from './types.js'
import type {
  Color,
  KnownMaterial,
  Link,
  Material,
  Robot,
  Visual,
} from './urdf.js'
import { getRootLink } from './urdf.js'
import { compact, isString, isNotEmpty, size } from './utils.js'
import {
  defaultImport,
  namedImport,
  exportDefault,
  declareUseTexture,
  makeArrowFunction,
  makeJsxAttr,
  makeJsxElement,
  makeJsxLink,
  makeJsxJoint,
  makeParameter,
  removeExtraClosingTags,
  uniqueComponents,
} from './ast.js'
import {
  defineResolveFunction,
  defineLinkProps,
  defineLinkComponent,
  defineJointProps,
  defineJointComponent,
  defineModelProps,
  defineModelComponent,
} from './stub.js'

const TEXTURE_PREFIX = '_texture'

const ALL_IMPORTS: [string, string[]][] = [
  ['three/examples/jsm/loaders/STLLoader', ['STLLoader']],
  ['@react-three/fiber', ['useLoader', 'GroupProps', 'MeshProps']],
  ['@react-three/drei', ['useTexture', 'Box', 'Cylinder', 'Sphere']],
]

const ALL_DECLARATIONS: [string, t.Statement][] = [
  ['defineResolveFunction', defineResolveFunction],
  ['defineLinkProps', defineLinkProps],
  ['defineLinkComponent', defineLinkComponent],
  ['defineJointProps', defineJointProps],
  ['defineJointComponent', defineJointComponent],
  ['defineModelProps', defineModelProps],
  ['defineModelComponent', defineModelComponent],
]

const DEPENDENCY: Dict<string[]> = {
  useTexture: ['useTexture', 'defineResolveFunction'],
  group: ['GroupProps'],
  Box: ['Box'],
  Cylinder: ['Cylinder'],
  Sphere: ['Sphere'],
  Link: ['GroupProps', 'defineLinkProps', 'defineLinkComponent'],
  Joint: ['GroupProps', 'defineJointProps', 'defineJointComponent'],
  Model: [
    'MeshProps',
    'useLoader',
    'STLLoader',
    'defineResolveFunction',
    'defineModelProps',
    'defineModelComponent',
  ],
}

const dependencies = (targets: string[]) => {
  const lookupDependency = (target: string) => DEPENDENCY[target] || []
  return R.uniq(R.flatten(R.map(lookupDependency, targets)))
}

const makeImportStatements = (deps: string[]) =>
  ALL_IMPORTS.reduce((statements, [module, allNames]) => {
    const names = R.intersection(allNames, deps)
    return R.isEmpty(names)
      ? statements
      : [...statements, namedImport(R.sortBy(R.identity, names), module)]
  }, [] as t.Statement[])

const makeDeclarationStatements = (deps: string[]) =>
  ALL_DECLARATIONS.reduce(
    (statements, [name, statement]) =>
      deps.includes(name) ? [...statements, statement] : statements,
    [] as t.Statement[]
  )

// serve as placeholder for blank line
export const withLeadingBlankComment = <T extends t.Node>(node: T): T => ({
  ...node,
  leadingComments: [{ type: 'CommentLine', value: '' }],
})

export const removeEmptyComments = (code: string) =>
  code.replace(/\/\/\s*\n/g, '\n')

const setMaterial = (visual: Visual, nameOrColor: string | Color): Visual => {
  const material: KnownMaterial = isString(nameOrColor)
    ? { name: nameOrColor }
    : { color: nameOrColor }
  return { ...visual, material }
}

const populateMaterial = (
  links: Link[],
  globalMaterial: Dict<string | Color>
) => {
  type Result<T> = [T[], Dict<string>]

  const swapMaterial = (visuals: Visual[], localTexture: Dict<string>) =>
    visuals.reduce(
      ([updatedVisuals, localTexture], visual) => {
        const addToResult = (visual: Visual, newTexture?: Dict<string>) =>
          [
            [...updatedVisuals, visual],
            { ...localTexture, ...newTexture },
          ] as Result<Visual>

        if (R.anyPass([R.isNil, R.isEmpty])(visual.material)) {
          return addToResult(visual)
        }

        const { name, color, texture } = visual.material!
        if (name) {
          const { [name]: textureOrColor } = globalMaterial
          if (!textureOrColor) {
            throw new Error(`Unknown material: ${name}`)
          }

          const nameOrColor = isString(textureOrColor) ? name : textureOrColor
          return addToResult(setMaterial(visual, nameOrColor))
        }
        if (color) {
          return addToResult(setMaterial(visual, color))
        }

        const textureName = `${TEXTURE_PREFIX}${size(localTexture)}`
        const newTexture = { [textureName]: texture! }
        return addToResult(setMaterial(visual, textureName), newTexture)
      },
      [[], localTexture] as Result<Visual>
    )

  return links.reduce(
    ([updatedLinks, texture], link) => {
      const [visual, localTexture] = swapMaterial(link.visual, texture)
      const updatedLink: Link = { ...link, visual }
      return [[...updatedLinks, updatedLink], { ...texture, ...localTexture }]
    },
    [[], {}] as Result<Link>
  ) as [Link[], Dict<string>]
}

const materialLookup = (materials: Material[]): Dict<string | Color> =>
  Object.fromEntries(
    R.map(({ name, color, texture }) => [name!, (color || texture)!], materials)
  )

export const transform = (robot: Robot) => {
  const globalMaterial = materialLookup(robot.material)
  const [links, localTexture] = populateMaterial(robot.link, globalMaterial)

  const globalTexture = R.filter(isString, globalMaterial)
  const texture = R.mergeAll([globalTexture, localTexture])

  const jsxLinksByName = Object.fromEntries(
    R.map((link) => [link.name, makeJsxLink(link)], links)
  )

  const jsxJoints: t.JSXElement[] = []
  for (const joint of robot.joint) {
    const { parent, child } = joint
    const jsxJoint = makeJsxJoint(joint)
    jsxLinksByName[parent].children.push(jsxJoint)
    jsxJoint.children.push(jsxLinksByName[child])
    jsxJoints.push(jsxJoint)
  }

  const jsxRobot = makeJsxElement(
    'group',
    [
      makeJsxAttr('name', robot.name),
      t.jsxSpreadAttribute(t.identifier('props')),
    ],
    [jsxLinksByName[getRootLink(robot)]]
  )

  const hasTexture = isNotEmpty(texture)

  const deps = dependencies(
    compact([hasTexture && 'useTexture', ...uniqueComponents(jsxRobot)])
  )

  const defineRobotComponent = makeArrowFunction(
    'Robot',
    [makeParameter('props', 'GroupProps')],
    hasTexture
      ? t.blockStatement([
          declareUseTexture(texture),
          t.returnStatement(jsxRobot),
        ])
      : jsxRobot
  )

  const program = t.program([
    defaultImport('React', 'react'),
    ...makeImportStatements(deps),
    ...R.map(withLeadingBlankComment, makeDeclarationStatements(deps)),
    withLeadingBlankComment(defineRobotComponent),
    withLeadingBlankComment(exportDefault('Robot')),
  ])
  return removeExtraClosingTags(program)
}
