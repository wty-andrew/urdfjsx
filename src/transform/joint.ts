import * as t from '@babel/types'
import * as R from 'ramda'

import { makeAnnotatedIdentifier, makeObjectProperty } from '../ast/index.js'
import type { Joint, Limit, Vector3, Vector4 } from '../types/index.js'
import { eulerToQuaternion } from '../utils/index.js'

const makeThreeObj = (type: string, ...args: number[]) =>
  t.newExpression(
    t.memberExpression(t.identifier('THREE'), t.identifier(type)),
    R.map<number, t.NumericLiteral>(t.numericLiteral, args)
  )

const makeThreeVector3 = (vec: Vector3) => makeThreeObj('Vector3', ...vec)

const makeThreeQuaternion = (quaternion: Vector4) =>
  makeThreeObj('Quaternion', ...quaternion)

const makeLimitProperties = ({ lower, upper }: Limit) => [
  makeObjectProperty('lower', lower),
  makeObjectProperty('upper', upper),
]

const makeJointProperties = ({
  type,
  axis = [1, 0, 0],
  limit,
  origin,
}: Joint) => {
  switch (type) {
    case 'fixed':
    case 'floating':
    case 'planar':
      return [makeObjectProperty('type', type)]
    case 'continuous':
    case 'revolute': {
      const offset = eulerToQuaternion(origin?.rpy || [0, 0, 0], 'ZYX')
      return [
        makeObjectProperty('type', type),
        makeObjectProperty('axis', makeThreeVector3(axis)),
        makeObjectProperty('offset', makeThreeQuaternion(offset)),
        ...(type === 'revolute' ? makeLimitProperties(limit!) : []),
      ]
    }
    case 'prismatic': {
      const offset = origin?.xyz || [0, 0, 0]
      return [
        makeObjectProperty('type', type),
        makeObjectProperty('axis', makeThreeVector3(axis)),
        makeObjectProperty('offset', makeThreeVector3(offset)),
        ...makeLimitProperties(limit!),
      ]
    }
  }
}

const makeJointSchema = (joint: Joint) =>
  makeObjectProperty(joint.name, t.objectExpression(makeJointProperties(joint)))

export const declareJointSchema = (joints: Joint[]) => {
  const annotation = t.tsTypeReference(
    t.identifier('Record'),
    t.tsTypeParameterInstantiation([
      t.tsStringKeyword(),
      t.tsTypeReference(t.identifier('JointSchema')),
    ])
  )
  return t.exportNamedDeclaration(
    t.variableDeclaration('const', [
      t.variableDeclarator(
        makeAnnotatedIdentifier('jointSchema', annotation),
        t.objectExpression(
          R.map<Joint, t.ObjectProperty>(makeJointSchema, joints)
        )
      ),
    ])
  )
}
