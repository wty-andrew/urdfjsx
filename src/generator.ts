import * as t from '@babel/types'
import ts from 'typescript'

import { formatAST, prettify } from './ast.js'
import { removeEmptyComments } from './transform.js'

export enum Target {
  TSX = 'tsx',
  JSX = 'jsx',
}

const transpile = (code: string, options?: ts.CompilerOptions) =>
  ts.transpileModule(code, {
    compilerOptions: {
      target: ts.ScriptTarget.ES2022,
      module: ts.ModuleKind.ES2022,
      jsx: ts.JsxEmit.Preserve,
      removeComments: false,
      ...options,
    },
  }).outputText

export const generateCode = async (progam: t.Program, target: Target) => {
  const tsx = formatAST(progam)
  const code = target === 'tsx' ? tsx : transpile(tsx)
  return await prettify(removeEmptyComments(code))
}
