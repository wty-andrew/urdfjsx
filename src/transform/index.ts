export * from './joint.js'
export * from './material.js'
export * from './robot.js'
export * from './statements.js'

import * as t from '@babel/types'
import * as R from 'ramda'

import type { Robot } from '../types/index.js'
import { compact, isString, size } from '../utils/index.js'
import { exportDefault, withLeadingBlankComment } from '../ast/index.js'
import { materialLookup, populateMaterial } from './material.js'
import { declareJointSchema } from './joint.js'
import { makeJsxRobot, uniqueComponents } from './robot.js'
import {
  dependencies,
  makeImportStatements,
  makeDeclarationStatements,
  defineJointSchema,
  defineRobotComponent,
} from './statements.js'

export const transform = (robot: Robot) => {
  const globalMaterial = materialLookup(robot.material)
  const globalTexture = R.filter(isString, globalMaterial)
  const [links, localTexture] = populateMaterial(robot.link, globalMaterial)
  const texture = R.mergeAll([globalTexture, localTexture])

  const jsxRobot = makeJsxRobot({ ...robot, link: links })

  const deps = dependencies(
    compact([size(texture) > 0 && 'useTexture', ...uniqueComponents(jsxRobot)])
  )

  return t.program([
    ...makeImportStatements(deps),
    ...R.map(withLeadingBlankComment, [
      ...makeDeclarationStatements(deps),
      defineJointSchema,
      declareJointSchema(robot.joint),
      defineRobotComponent(jsxRobot, texture),
      exportDefault('Robot'),
    ]),
  ])
}
