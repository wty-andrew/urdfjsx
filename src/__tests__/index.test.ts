import fs from 'node:fs'
import path from 'node:path'
import { fileURLToPath } from 'node:url'

import { parseURDF } from '../urdf.js'
import { transform } from '../transform.js'
import { generateCode, Target } from '../generator.js'

const __dirname = path.dirname(fileURLToPath(import.meta.url))

const loadFixture = (filename: string) =>
  fs.readFileSync(path.join(__dirname, `fixtures/${filename}`), 'utf-8')

const urdfToTsx = (urdf: string) =>
  generateCode(transform(parseURDF(urdf)), Target.TSX)

expect.addSnapshotSerializer({
  test: (value) => typeof value === 'string',
  print: (value) => (value as string).trim(),
})

test('URDF to TSX conversion', async () => {
  const tsx = await urdfToTsx(loadFixture('urdf-tutorial.urdf'))
  expect(tsx).toMatchSnapshot()
})
