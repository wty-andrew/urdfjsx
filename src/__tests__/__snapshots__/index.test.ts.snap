// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`URDF to TSX conversion 1`] = `
import React from 'react'
import { STLLoader } from 'three/examples/jsm/loaders/STLLoader'
import { GroupProps, MeshProps, useLoader } from '@react-three/fiber'
import { Box, Cylinder, Sphere } from '@react-three/drei'

const resolve = (url: string) => url

interface LinkProps extends Omit<GroupProps, 'position' | 'rotation'> {
  name: string
}

const Link = ({ name, ...props }: LinkProps) => <group name={name} {...props} />

interface JointProps extends GroupProps {
  name: string
}

const Joint = ({ name, ...props }: JointProps) => (
  <group name={name} {...props} />
)

interface ModelProps extends MeshProps {
  url: string
}

const Model = ({ url, children, ...props }: ModelProps) => {
  const geometry = useLoader(STLLoader, url)
  return (
    <mesh geometry={geometry} {...props}>
      {children}
    </mesh>
  )
}

const Robot = (props: GroupProps) => (
  <group name="visual" {...props}>
    <Link name="base_link">
      <Cylinder args={[0.2, 0.2, 0.6]}>
        <meshStandardMaterial color={[0, 0, 0.8]} />
      </Cylinder>
      <Joint name="base_to_right_leg" position={[0, -0.22, 0.25]}>
        <Link name="right_leg">
          <Box
            args={[0.6, 0.1, 0.2]}
            position={[0, 0, -0.3]}
            rotation={[0, 1.57075, 0, 'ZYX']}
          >
            <meshStandardMaterial color={[1, 1, 1]} />
          </Box>
          <Joint name="right_base_joint" position={[0, 0, -0.6]}>
            <Link name="right_base">
              <Box args={[0.4, 0.1, 0.1]}>
                <meshStandardMaterial color={[1, 1, 1]} />
              </Box>
              <Joint
                name="right_front_wheel_joint"
                position={[0.133333333333, 0, -0.085]}
                rotation={[0, 0, 0, 'ZYX']}
              >
                <Link name="right_front_wheel">
                  <Cylinder
                    args={[0.035, 0.035, 0.1]}
                    position={[0, 0, 0]}
                    rotation={[1.57075, 0, 0, 'ZYX']}
                  >
                    <meshStandardMaterial color={[0, 0, 0]} />
                  </Cylinder>
                </Link>
              </Joint>
              <Joint
                name="right_back_wheel_joint"
                position={[-0.133333333333, 0, -0.085]}
                rotation={[0, 0, 0, 'ZYX']}
              >
                <Link name="right_back_wheel">
                  <Cylinder
                    args={[0.035, 0.035, 0.1]}
                    position={[0, 0, 0]}
                    rotation={[1.57075, 0, 0, 'ZYX']}
                  >
                    <meshStandardMaterial color={[0, 0, 0]} />
                  </Cylinder>
                </Link>
              </Joint>
            </Link>
          </Joint>
        </Link>
      </Joint>
      <Joint name="base_to_left_leg" position={[0, 0.22, 0.25]}>
        <Link name="left_leg">
          <Box
            args={[0.6, 0.1, 0.2]}
            position={[0, 0, -0.3]}
            rotation={[0, 1.57075, 0, 'ZYX']}
          >
            <meshStandardMaterial color={[1, 1, 1]} />
          </Box>
          <Joint name="left_base_joint" position={[0, 0, -0.6]}>
            <Link name="left_base">
              <Box args={[0.4, 0.1, 0.1]}>
                <meshStandardMaterial color={[1, 1, 1]} />
              </Box>
              <Joint
                name="left_front_wheel_joint"
                position={[0.133333333333, 0, -0.085]}
                rotation={[0, 0, 0, 'ZYX']}
              >
                <Link name="left_front_wheel">
                  <Cylinder
                    args={[0.035, 0.035, 0.1]}
                    position={[0, 0, 0]}
                    rotation={[1.57075, 0, 0, 'ZYX']}
                  >
                    <meshStandardMaterial color={[0, 0, 0]} />
                  </Cylinder>
                </Link>
              </Joint>
              <Joint
                name="left_back_wheel_joint"
                position={[-0.133333333333, 0, -0.085]}
                rotation={[0, 0, 0, 'ZYX']}
              >
                <Link name="left_back_wheel">
                  <Cylinder
                    args={[0.035, 0.035, 0.1]}
                    position={[0, 0, 0]}
                    rotation={[1.57075, 0, 0, 'ZYX']}
                  >
                    <meshStandardMaterial color={[0, 0, 0]} />
                  </Cylinder>
                </Link>
              </Joint>
            </Link>
          </Joint>
        </Link>
      </Joint>
      <Joint
        name="gripper_extension"
        position={[0.19, 0, 0.2]}
        rotation={[0, 0, 0, 'ZYX']}
      >
        <Link name="gripper_pole">
          <Cylinder
            args={[0.01, 0.01, 0.2]}
            position={[0.1, 0, 0]}
            rotation={[0, 1.57075, 0, 'ZYX']}
          >
            <meshStandardMaterial color={[1, 1, 1]} />
          </Cylinder>
          <Joint
            name="left_gripper_joint"
            position={[0.2, 0.01, 0]}
            rotation={[0, 0, 0, 'ZYX']}
          >
            <Link name="left_gripper">
              <Model
                url={resolve('/urdf_tutorial/meshes/l_finger.dae')}
                position={[0, 0, 0]}
                rotation={[0, 0, 0, 'ZYX']}
              >
                <meshStandardMaterial color={[1, 1, 1]} />
              </Model>
              <Joint name="left_tip_joint">
                <Link name="left_tip">
                  <Model
                    url={resolve('/urdf_tutorial/meshes/l_finger_tip.dae')}
                    position={[0.09137, 0.00495, 0]}
                    rotation={[0, 0, 0, 'ZYX']}
                  >
                    <meshStandardMaterial color={[1, 1, 1]} />
                  </Model>
                </Link>
              </Joint>
            </Link>
          </Joint>
          <Joint
            name="right_gripper_joint"
            position={[0.2, -0.01, 0]}
            rotation={[0, 0, 0, 'ZYX']}
          >
            <Link name="right_gripper">
              <Model
                url={resolve('/urdf_tutorial/meshes/l_finger.dae')}
                position={[0, 0, 0]}
                rotation={[-3.1415, 0, 0, 'ZYX']}
              >
                <meshStandardMaterial color={[1, 1, 1]} />
              </Model>
              <Joint name="right_tip_joint">
                <Link name="right_tip">
                  <Model
                    url={resolve('/urdf_tutorial/meshes/l_finger_tip.dae')}
                    position={[0.09137, 0.00495, 0]}
                    rotation={[-3.1415, 0, 0, 'ZYX']}
                  >
                    <meshStandardMaterial color={[1, 1, 1]} />
                  </Model>
                </Link>
              </Joint>
            </Link>
          </Joint>
        </Link>
      </Joint>
      <Joint name="head_swivel" position={[0, 0, 0.3]}>
        <Link name="head">
          <Sphere args={[0.2]}>
            <meshStandardMaterial color={[1, 1, 1]} />
          </Sphere>
          <Joint name="tobox" position={[0.1814, 0, 0.1414]}>
            <Link name="box">
              <Box args={[0.08, 0.08, 0.08]}>
                <meshStandardMaterial color={[0, 0, 0.8]} />
              </Box>
            </Link>
          </Joint>
        </Link>
      </Joint>
    </Link>
  </group>
)

export default Robot
`;
